zkevm-circuits/src/rlp_circuit_fsm.rs

```diff
@@ -1405,12 +1497,489 @@ impl<F: Field> RlpCircuitConfig<F> {
+            |meta| {
+                let enable = and::expr([
+                    meta.query_fixed(q_enabled, Rotation::cur()),
+                    meta.query_advice(rlp_decoding_table.is_stack_push, Rotation::cur()),
+                ]);
+
+                let input_exprs = vec![
+                    meta.query_advice(rlp_decoding_table.id, Rotation::cur()),
+                    meta.query_advice(rlp_decoding_table.depth, Rotation::cur()) - 1.expr(),
+                    1.expr(), // is_pop = true
+                    meta.query_advice(rlp_decoding_table.value, Rotation::cur()) + 1.expr(),
+                ];
+                let table_exprs = vec![
+                    meta.query_advice(rlp_decoding_table.id, Rotation::cur()),
+                    meta.query_advice(rlp_decoding_table.depth, Rotation::cur()),
+                    meta.query_advice(rlp_decoding_table.is_stack_pop, Rotation::cur()),
+                    meta.query_advice(rlp_decoding_table.value_prev, Rotation::cur())
+                        - meta.query_advice(rlp_decoding_table.value, Rotation::cur()),
+                ];
+                input_exprs
+                    .into_iter()
+                    .zip(table_exprs)
+                    .map(|(input, table)| (input * enable.expr(), table))
+                    .collect()
+            },
+        );
+        meta.lookup_any(
+            "Each stack POP must correspond exactly to a PUSH into a higher depth level",
+            |meta| {
+                let enable = and::expr([
+                    meta.query_fixed(q_enabled, Rotation::cur()),
+                    meta.query_advice(rlp_decoding_table.is_stack_pop, Rotation::cur()),
+                ]);
+
+                let input_exprs = vec![
+                    meta.query_advice(rlp_decoding_table.id, Rotation::cur()),
+                    meta.query_advice(rlp_decoding_table.depth, Rotation::cur()) + 1.expr(),
+                    1.expr(), // is_push = true
+                    meta.query_advice(rlp_decoding_table.value_prev, Rotation::cur())
+                        - meta.query_advice(rlp_decoding_table.value, Rotation::cur())
+                        - 1.expr(),
+                    0.expr(), // value_prev = 0
+                ];
+                let table_exprs = vec![
+                    meta.query_advice(rlp_decoding_table.id, Rotation::cur()),
+                    meta.query_advice(rlp_decoding_table.depth, Rotation::cur()),
+                    meta.query_advice(rlp_decoding_table.is_stack_push, Rotation::cur()),
+                    meta.query_advice(rlp_decoding_table.value, Rotation::cur()),
+                    meta.query_advice(rlp_decoding_table.value_prev, Rotation::cur()),
+                ];
+                input_exprs
+                    .into_iter()
+                    .zip(table_exprs)
+                    .map(|(input, table)| (input * enable.expr(), table))
+                    .collect()
+            },
+        );
```

Comment:
Why do we need so many (5) lookups for rlp decoding table? I think 2 lookups is enough. We only need to ensure the following 5 types of state transition for RlpDecodingTable:
1. We group the rows of rlp decoding table by `key = (tx_id, format, depth, al_idx, sk_idx)`, and `byte_idx` is used for sorting rows in same group;
2. Whenever the key changes (i.e. `key::prev != key::cur`), we asserts that cur row is the first row in the group with `key::cur`, i.e. `value_prev::cur = 0` and `value` is constrained by the sm. This corresponds to `PUSH`.
3. Whenever the key changes (i.e. `key::prev != key::cur`), we asserts that prev row is the last row in the group with `key::prev`, i.e. `value::prev = 0`. This corresponds to `POP`.
4. When key does not change (i.e. `key::prev = key::cur`), we asserts that `value_prev::cur = value::prev` and `value::cur + 1 = value_prev::cur`. This corresponds to `UPDATE`.
5. Keys are mononically increasing.

Reply:
For case 2, it's not necessarily the case that a key change corresponds to a `PUSH` operation. Consider an access list item with a storage key list that has multiple keys, for example:

```
...
{
    "address": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
    "storageKeys": [
      "0xe1dd9768c9de657aca2536cf1cdd1c4536b13ec81ff764307ea8312aa7a8790d",
      "0x70bc879403c8b875e45ea7afbb591f1fd4bde469db47d5f0e879e44c6798d33e",
      "0xf88aa3ad276c350a067c34b2bed705e1a2cd30c7c3154f62ece8ee00939bbd2e",
      "0xbe11b0e2ba48478671bfcd8fd182e025c26fbfbcf4fdf6952051d6147955a36f",
      "0x9a1a5a7ef77f3399dea2a1044425aaca7fec294fdfdcacd7a960c9c94d15f0a6",
      "0x91828b9b711948523369ff1651b6332e98f75bcd940a551dc7247d5af88e71fa"
    ]
},
...
```

When the second storage key is processed, the `sk_idx` increments, but no new `PUSH` onto level 4 is present as `PUSH` only happens with `BeginVector` at the start of the storage key list. In this case, there's only 1 `PUSH` op with all bytes for all 6 storage keys pushed onto depth 4. This pattern is observed in the `state_machine`.  For all storage keys except the first one, the key change will actually correspond to an `UPDATE` operation. 

There're some other nuances when it comes to key change with the witness value (the stack operations). These scenarios are explained below. When key is set to be `(tx_id, format, depth, al_idx, sk_idx)`, here're all the scenarios of key change and key consistency for sorted stack operations:

**A. When key changes**
- `PUSH`:
1. `depth=1`: start decoding after `TxType` and `Len` on depth 0
2. `depth=2`: start decoding the access list
3. `depth=3`: a new access list item
4. `depth=4`: a new storage key

- `POP`:
1. `depth=0`: finished decoding entire RLP bytes
3. `depth=2`: finished decoding an access list item

- `UPDATE`:
1. `depth=4`: a new access list storage key

**B. When key stays the same**
- `POP`:
2. `depth=1`: finished decoding the access list
3. `depth=3`: finished decoding the storage key list for an access list item

- `UPDATE`: read a regular byte on the same depth level

To help better communicate and understand on this branch, I've prepared a full manifest of the witness values for the RLP circuit, divided into `rlp_table`, `state_machine` and `rlp_decoding_table` sections. The stack ops in their sorted state are also included. I've sent you a link to this document.